File: servers/typescript/src/pool_broadcast.ts
import type { StarfishFrame } from "./types.js";
import type { StarfishServer } from "./starfish_server.js";
import type { Client } from "./client.js";
import type { Pool, MatchResult } from "./pool.js";
import { createErrorFrame, ERR_POOL_NOT_FOUND, ERR_POOL_NOT_MEMBER } from "./errors.js";

export function resolvePool(hub: StarfishServer, client: Client, frame: StarfishFrame): Pool | undefined {
  const payload = frame.payload as { pool?: string } | undefined;
  const pool = payload?.pool ? hub.getPool(payload.pool) : undefined;
  if (!pool) {
    client.sendFrame(createErrorFrame(hub.idGen, frame.header.id, ERR_POOL_NOT_FOUND, "pool", frame.header.method));
  }
  return pool;
}

export function requirePoolMember(pool: Pool, hub: StarfishServer, client: Client, frame: StarfishFrame): boolean {
  if (!pool.hasMember(client.id)) {
    client.sendFrame(createErrorFrame(hub.idGen, frame.header.id, ERR_POOL_NOT_MEMBER, "pool", frame.header.method));
    return false;
  }
  return true;
}

export function broadcastMemberJoined(
  hub: StarfishServer,
  pool: Pool,
  newMemberId: string,
  attributes: Record<string, unknown>,
): void {
  const frame: StarfishFrame = {
    header: {
      id: hub.idGen.messageId(),
      resource: "pool",
      method: "member-joined",
      kind: "event",
    },
    payload: { pool: pool.name, member: { id: newMemberId, attributes } },
  };

  if (pool.isClaimBased()) {
    for (const member of pool.getMembers()) {
      if (member.clientId !== newMemberId) {
        hub.getClient(member.clientId)?.sendFrame(frame);
      }
    }
  } else if (pool.mode === "delegated") {
    for (const member of pool.getMembers()) {
      if (member.clientId !== newMemberId && pool.isMatchmaker(member.clientId)) {
        hub.getClient(member.clientId)?.sendFrame(frame);
      }
    }
  }
}

export function broadcastMemberLeft(
  hub: StarfishServer,
  pool: Pool,
  memberId: string,
  reason: string,
): void {
  const frame: StarfishFrame = {
    header: {
      id: hub.idGen.messageId(),
      resource: "pool",
      method: "member-left",
      kind: "event",
    },
    payload: { pool: pool.name, memberId, reason },
  };

  if (pool.isClaimBased()) {
    for (const member of pool.getMembers()) {
      if (member.clientId !== memberId) {
        hub.getClient(member.clientId)?.sendFrame(frame);
      }
    }
  } else if (pool.mode === "delegated") {
    for (const member of pool.getMembers()) {
      if (member.clientId !== memberId && pool.isMatchmaker(member.clientId)) {
        hub.getClient(member.clientId)?.sendFrame(frame);
      }
    }
  }
}

export function sendMatchedToGroup(
  hub: StarfishServer,
  result: MatchResult,
  poolName: string,
): void {
  for (const peer of result.peers) {
    hub.getClient(peer.id)?.sendFrame({
      header: {
        id: hub.idGen.messageId(),
        resource: "pool",
        method: "matched",
        kind: "event",
      },
      payload: { pool: poolName, session: result.session, peers: result.peers },
    });
  }
}

export function broadcastMatchedMembersLeft(
  hub: StarfishServer,
  pool: Pool,
  matchedIds: string[],
): void {
  if (!pool.isClaimBased()) return;

  for (const matchedId of matchedIds) {
    const frame: StarfishFrame = {
      header: {
        id: hub.idGen.messageId(),
        resource: "pool",
        method: "member-left",
        kind: "event",
      },
      payload: { pool: pool.name, memberId: matchedId, reason: "matched" },
    };
    for (const member of pool.getMembers()) {
      hub.getClient(member.clientId)?.sendFrame(frame);
    }
  }
}

File: servers/python/src/starfish_server/pool_broadcast.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .client import Client
from .errors import ERR_POOL_NOT_FOUND, ERR_POOL_NOT_MEMBER, create_error_frame
from .pool import MatchResult, Pool
from .types import StarfishFrame

if TYPE_CHECKING:
    from .server import StarfishServer


def resolve_pool(hub: StarfishServer, client: Client, frame: StarfishFrame) -> Pool | None:
    payload = frame.get("payload") or {}
    pool_name = payload.get("pool")
    pool = hub.get_pool(pool_name) if pool_name else None
    if pool is None:
        header = frame.get("header", {})
        client.send_frame(
            create_error_frame(hub.id_gen, header.get("id", ""), ERR_POOL_NOT_FOUND, "pool", header.get("method"))
        )
    return pool


def require_pool_member(pool: Pool, hub: StarfishServer, client: Client, frame: StarfishFrame) -> bool:
    if not pool.has_member(client.id):
        header = frame.get("header", {})
        client.send_frame(
            create_error_frame(hub.id_gen, header.get("id", ""), ERR_POOL_NOT_MEMBER, "pool", header.get("method"))
        )
        return False
    return True


def send_matched_to_group(hub: StarfishServer, result: MatchResult, pool_name: str) -> None:
    for peer in result.peers:
        c = hub.get_client(peer["id"])
        if c:
            c.send_frame({
                "header": {
                    "id": hub.id_gen.message_id(),
                    "resource": "pool",
                    "method": "matched",
                    "kind": "event",
                },
                "payload": {"pool": pool_name, "session": result.session, "peers": result.peers},
            })


def broadcast_matched_members_left(hub: StarfishServer, pool: Pool, matched_ids: list[str]) -> None:
    if not pool.is_claim_based():
        return
    for matched_id in matched_ids:
        frame: StarfishFrame = {
            "header": {
                "id": hub.id_gen.message_id(),
                "resource": "pool",
                "method": "member-left",
                "kind": "event",
            },
            "payload": {"pool": pool.name, "memberId": matched_id, "reason": "matched"},
        }
        for member in pool.get_members():
            c = hub.get_client(member.client_id)
            if c:
                c.send_frame(frame)


def broadcast_pool_member_joined(
    hub: StarfishServer,
    pool: Pool,
    client_id: str,
    attributes: dict,
) -> None:
    frame: StarfishFrame = {
        "header": {
            "id": hub.id_gen.message_id(),
            "resource": "pool",
            "method": "member-joined",
            "kind": "event",
        },
        "payload": {
            "pool": pool.name,
            "member": {"id": client_id, "attributes": attributes},
        },
    }

    if pool.is_claim_based():
        for member in pool.get_members():
            if member.client_id != client_id:
                c = hub.get_client(member.client_id)
                if c:
                    c.send_frame(frame)
    elif pool.mode == "delegated":
        for member in pool.get_members():
            if member.client_id != client_id and pool.is_matchmaker(member.client_id):
                c = hub.get_client(member.client_id)
                if c:
                    c.send_frame(frame)


def broadcast_pool_member_left(
    hub: StarfishServer,
    pool: Pool,
    member_id: str,
    reason: str,
) -> None:
    frame: StarfishFrame = {
        "header": {
            "id": hub.id_gen.message_id(),
            "resource": "pool",
            "method": "member-left",
            "kind": "event",
        },
        "payload": {"pool": pool.name, "memberId": member_id, "reason": reason},
    }

    if pool.is_claim_based():
        for member in pool.get_members():
            if member.client_id != member_id:
                c = hub.get_client(member.client_id)
                if c:
                    c.send_frame(frame)
    elif pool.mode == "delegated":
        for member in pool.get_members():
            if member.client_id != member_id and pool.is_matchmaker(member.client_id):
                c = hub.get_client(member.client_id)
                if c:
                    c.send_frame(frame)
